blocks/monitoring/aggregation.py

```diff
@@ -138,11 +138,62 @@ def get_aggregator(self):
         return aggregator
 
 
-def mean(numerator, denominator=1.):
+def mean(numerator, denominator=1., name=None):
     """Mean of quantity (numerator) over a number (denominator) values."""
     variable = numerator / denominator
     variable.tag.aggregation_scheme = Mean(numerator, denominator)
-    variable.name = numerator.name
+    if name is not None:
```

Comment:
I know about TrainingDataMonitoring but the problem is the same - you 
want to pass a named variable.

I was thinking about fixing named_copy but I guess it will be too error 
prone - we either stay compatible with variable.copy() which removes 
annotations, or we will introduce very subtle bugs - trust me, It took 
me half a day to figure that I lost my aggregation schemes and I 
discovered it only because I didn't rely on the monitoring to do the 
final scoring.

So we either work with Theano guys making sure that normal copies made 
by theano preserve the annotations, or we name the variables in a 
convenient way and discourage making copies.

Dmitry Bogdanov wrote:
>
> In blocks/monitoring/aggregation.py
> <https://github.com/mila-udem/blocks/pull/780#discussion_r36044701>:
>
> >       """Mean of quantity (numerator) over a number (denominator) 
> values."""
> >       variable = numerator / denominator
> >       variable.tag.aggregation_scheme = Mean(numerator, denominator)
> >  -    variable.name = numerator.name
> >  +    if name is not None:
>
> We don't really have |extension.monitor(...)| in Blocks now. One
> creates |TrainingDataMonitoring| or |DataStreamMonitoring| with the
> list of all variables to be monitored.
>
> There is also a homogeneity argument: if you create a variable for
> monitoring without adding an aggregation scheme, you have to give it a
> name in a separate line.
>
> Maybe we should make sure that |named_copy| doesn't lose the
> aggregation scheme and whatever else it loses?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/780/files#r36044701>.


> Dmitry Bogdanov <mailto:notifications@github.com>
> Sunday, August 02, 2015 6:28 PM
>
> In blocks/monitoring/aggregation.py 
> <https://github.com/mila-udem/blocks/pull/780#discussion_r36044701>:
>
> > """Mean of quantity (numerator) over a number (denominator) values."""
> > variable = numerator / denominator
> > variable.tag.aggregation_scheme = Mean(numerator, denominator)
> > - variable.name = numerator.name
> > + if name is not None:
>
> We don't really have |extension.monitor(...)| in Blocks now. One 
> creates |TrainingDataMonitoring| or |DataStreamMonitoring| with the 
> list of all variables to be monitored.
>
> There is also a homogeneity argument: if you create a variable for 
> monitoring without adding an aggregation scheme, you have to give it a 
> name in a separate line.
>
> Maybe we should make sure that |named_copy| doesn't lose the 
> aggregation scheme and whatever else it loses?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mila-udem/blocks/pull/780/files#r36044701>.
>